File: sysup/internal/download/download.go
// Package download is the shared "fetch a URL, optionally verify a
// checksums.txt entry, extract a tar.gz" logic used by both the polkit
// setup flow (downloading sysup-worker from a GitHub release) and the
// GitKraken installer — previously two independent hand-rolled
// implementations of the same GET+extract steps.
package download

import (
	"archive/tar"
	"compress/gzip"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// Get fetches url and returns the full response body.
func Get(url string) ([]byte, error) {
	client := &http.Client{Timeout: 60 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("GET %s: status %s", url, resp.Status)
	}
	return io.ReadAll(resp.Body)
}

// GetToFile fetches url and writes the response body straight to dest.
func GetToFile(url, dest string) error {
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("GET %s: status %s", url, resp.Status)
	}
	f, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = io.Copy(f, resp.Body)
	return err
}

// VerifyChecksum checks data's sha256 against the entry for assetName in a
// `sha256  filename` formatted checksums.txt (the format GoReleaser emits).
func VerifyChecksum(checksumsTxt, assetName string, data []byte) error {
	var want string
	for _, line := range strings.Split(checksumsTxt, "\n") {
		fields := strings.Fields(line)
		if len(fields) == 2 && fields[1] == assetName {
			want = fields[0]
			break
		}
	}
	if want == "" {
		return fmt.Errorf("checksum de %s não encontrado em checksums.txt", assetName)
	}
	sum := sha256.Sum256(data)
	got := hex.EncodeToString(sum[:])
	if got != want {
		return fmt.Errorf("checksum de %s não bate (esperado %s, obtido %s)", assetName, want, got)
	}
	return nil
}

// ExtractSingleFile pulls just wantName out of a tar.gz held in memory and
// writes it to a fresh temp file (0755), returning that path — used when
// only one binary inside the archive is actually needed (sysup-worker).
func ExtractSingleFile(tgz []byte, wantName string) (string, error) {
	gz, err := gzip.NewReader(strings.NewReader(string(tgz)))
	if err != nil {
		return "", err
	}
	defer gz.Close()

	tr := tar.NewReader(gz)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return "", err
		}
		if filepath.Base(hdr.Name) != wantName {
			continue
		}
		out := filepath.Join(os.TempDir(), fmt.Sprintf("%s-%d", wantName, os.Getpid()))
		f, err := os.OpenFile(out, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o755)
		if err != nil {
			return "", err
		}
		if _, err := io.Copy(f, tr); err != nil {
			f.Close()
			return "", err
		}
		f.Close()
		return out, nil
	}
	return "", fmt.Errorf("%s não encontrado no arquivo baixado", wantName)
}

// ExtractTarGz extracts every file/dir/symlink in the archive at
// archivePath into destDir, preserving structure and file modes — used
// when the whole tree is needed (GitKraken's install layout), unlike
// ExtractSingleFile.
func ExtractTarGz(archivePath, destDir string) error {
	f, err := os.Open(archivePath)
	if err != nil {
		return err
	}
	defer f.Close()

	gz, err := gzip.NewReader(f)
	if err != nil {
		return err
	}
	defer gz.Close()

	tr := tar.NewReader(gz)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		target := filepath.Join(destDir, hdr.Name)

		switch hdr.Typeflag {
		case tar.TypeDir:
			if err := os.MkdirAll(target, 0o755); err != nil {
				return err
			}
		case tar.TypeReg:
			if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
				return err
			}
			out, err := os.OpenFile(target, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.FileMode(hdr.Mode))
			if err != nil {
				return err
			}
			if _, err := io.Copy(out, tr); err != nil {
				out.Close()
				return err
			}
			out.Close()
		case tar.TypeSymlink:
			if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
				return err
			}
			_ = os.Remove(target)
			if err := os.Symlink(hdr.Linkname, target); err != nil {
				return err
			}
		}
	}
}

File: sysup/sysup/src/download.rs
use anyhow::{anyhow, Context};
use sha2::{Digest, Sha256};
use std::fs::File;
use std::io::Write;
use std::os::unix::fs::OpenOptionsExt;
use std::path::{Path, PathBuf};
use std::time::Duration;

pub fn get(url: &str) -> anyhow::Result<Vec<u8>> {
    let client = reqwest::blocking::Client::builder()
        .timeout(Duration::from_secs(60))
        .build()?;
    let resp = client.get(url).send()?;
    if !resp.status().is_success() {
        return Err(anyhow!("GET {}: status {}", url, resp.status()));
    }
    Ok(resp.bytes()?.to_vec())
}

pub fn get_to_file(url: &str, dest: &Path) -> anyhow::Result<()> {
    let resp = reqwest::blocking::get(url)?;
    if !resp.status().is_success() {
        return Err(anyhow!("GET {}: status {}", url, resp.status()));
    }
    let bytes = resp.bytes()?;
    let mut f = File::create(dest)?;
    f.write_all(&bytes)?;
    Ok(())
}

pub fn verify_checksum(checksums_txt: &str, asset_name: &str, data: &[u8]) -> anyhow::Result<()> {
    let mut want: Option<&str> = None;
    for line in checksums_txt.split('\n') {
        let fields: Vec<&str> = line.split_whitespace().collect();
        if fields.len() == 2 && fields[1] == asset_name {
            want = Some(fields[0]);
            break;
        }
    }
    let want = want.ok_or_else(|| {
        anyhow!("checksum de {} não encontrado em checksums.txt", asset_name)
    })?;

    let mut hasher = Sha256::new();
    hasher.update(data);
    let got = hasher
        .finalize()
        .iter()
        .map(|b| format!("{:02x}", b))
        .collect::<String>();
    if got != want {
        return Err(anyhow!(
            "checksum de {} não bate (esperado {}, obtido {})",
            asset_name,
            want,
            got
        ));
    }
    Ok(())
}

pub fn extract_single_file(tgz: &[u8], want_name: &str) -> anyhow::Result<PathBuf> {
    let gz = flate2::read::GzDecoder::new(tgz);
    let mut tr = tar::Archive::new(gz);

    for entry in tr.entries()? {
        let mut entry = entry?;
        let path = entry.path()?.into_owned();
        let name = match path.file_name() {
            Some(n) => n.to_string_lossy().into_owned(),
            None => continue,
        };
        if name != want_name {
            continue;
        }
        let out = std::env::temp_dir().join(format!("{}-{}", want_name, std::process::id()));
        let mut f = open_with_mode(&out, 0o755)?;
        std::io::copy(&mut entry, &mut f)?;
        return Ok(out);
    }
    Err(anyhow!("{} não encontrado no arquivo baixado", want_name))
}

pub fn extract_tar_gz(archive_path: &Path, dest_dir: &Path) -> anyhow::Result<()> {
    let f = File::open(archive_path)?;
    let gz = flate2::read::GzDecoder::new(f);
    let mut tr = tar::Archive::new(gz);

    for entry in tr.entries()? {
        let mut entry = entry?;
        let header = entry.header().clone();
        let name = entry.path()?.into_owned();
        let target = dest_dir.join(&name);

        match header.entry_type() {
            tar::EntryType::Directory => {
                std::fs::create_dir_all(&target)?;
            }
            tar::EntryType::Regular => {
                if let Some(parent) = target.parent() {
                    std::fs::create_dir_all(parent)?;
                }
                let mode = header.mode().unwrap_or(0o644);
                let mut out = open_with_mode(&target, mode)?;
                std::io::copy(&mut entry, &mut out)?;
            }
            tar::EntryType::Symlink => {
                if let Some(parent) = target.parent() {
                    std::fs::create_dir_all(parent)?;
                }
                let _ = std::fs::remove_file(&target);
                if let Some(link_name) = entry.link_name()? {
                    std::os::unix::fs::symlink(link_name, &target)
                        .with_context(|| format!("symlink {}", target.display()))?;
                }
            }
            _ => {}
        }
    }
    Ok(())
}

fn open_with_mode(path: &Path, mode: u32) -> anyhow::Result<File> {
    let f = std::fs::OpenOptions::new()
        .create(true)
        .write(true)
        .truncate(true)
        .mode(mode)
        .open(path)?;
    Ok(f)
}
